cmd/collector/app/span_processor.go

```diff
@@ -180,23 +182,55 @@ func (sp *spanProcessor) processItemFromQueue(item *queueItem) {
 	sp.metrics.InQueueLatency.Record(time.Since(item.queuedTime))
 }
 
-func (sp *spanProcessor) addCollectorTags(span *model.Span) {
-	if len(sp.collectorTags) == 0 {
-		return
-	}
-	dedupKey := make(map[string]struct{})
-	for _, tag := range span.Process.Tags {
-		if value, ok := sp.collectorTags[tag.Key]; ok && value == tag.AsString() {
-			sp.logger.Debug("ignore collector process tags", zap.String("key", tag.Key), zap.String("value", value))
-			dedupKey[tag.Key] = struct{}{}
+func (sp *spanProcessor) dedupTags(span *model.Span) {
+	deDupKey := make(map[string]int, len(span.Process.Tags))
+	deDupKeyVal := make(map[string]struct{}, len(span.Process.Tags))
+	for idx, tag := range span.Process.Tags {
+		if value, ok := sp.collectorTags[tag.Key]; ok {
+			if value == tag.AsString() {
+				deDupKeyVal[tag.Key] = struct{}{}
+			}
+			deDupKey[tag.Key] = idx
 		}
 	}
-	// ignore collector tags if has the same key-value in spans
-	for k, v := range sp.collectorTags {
-		if _, ok := dedupKey[k]; !ok {
-			span.Process.Tags = append(span.Process.Tags, model.String(k, v))
+
+	switch sp.useOnTagKeyConflict {
+	case "span":
+		// ignore collector tags if it has the same key in spans
+		// this will include the case of same key-value in spans too
+		for k, v := range sp.collectorTags {
+			if _, ok := deDupKey[k]; !ok {
+				span.Process.Tags = append(span.Process.Tags, model.String(k, v))
+			}
+		}
+	case "both":
+		// ignore collector tags if it has the same key-value in spans
+		for k, v := range sp.collectorTags {
+			if _, ok := deDupKeyVal[k]; !ok {
+				span.Process.Tags = append(span.Process.Tags, model.String(k, v))
+			}
 		}
+	case "collector":
+		// replace with collector tags if it has the same key in spans
+		for k, v := range sp.collectorTags {
+			if _, ok := deDupKeyVal[k]; !ok {
+				if _, ok = deDupKey[k]; ok {
+					idx := deDupKey[k]
+					copy(span.Process.Tags[idx:], span.Process.Tags[idx+1:])
+					span.Process.Tags = append(span.Process.Tags[:len(span.Process.Tags)-1], model.String(k, v))
```

Comment:
If I understand correctly, this is an O(n) time operation, together with a loop over the collector tags, _worst case_.

I think it's okay as it is, especially when _I think_ the modus operandi is not to set the collector tags and this case would only be hit if there are conflicts and the tag values are not identical.

Having said that, I was wondering if there is there a way we can simplify this?

I threw something together (which probably isn't perfect) to see if it helps us come up with ideas on how to simplify a bit. In particular, my goals were:
- Reduce the additional memory/complexity needed for maintaining additional maps
  - I still needed one additional map to store a copy of the collector tags.
- _Try_ to dedup the tags in a single pass over the process tags
  - I still needed one the additional loop over the collector tags at the end.
- Improve readability
  - Would like to get your thoughts on this.

Let me know what you think:
```
func (sp *spanProcessor) dedupTags(span *model.Span) {
	// Copy the collector tags to keep track of which of these need to be appended.
	cTags := make(map[string]string, len(sp.collectorTags))
	for k, v := range sp.collectorTags {
		cTags[k] = v
	}
	var procTags []model.KeyValue
	for _, sTag := range span.Process.Tags {
		appendSpanTag := true
		// If there's a tag key conflict.
		if cTagVal, ok := sp.collectorTags[sTag.Key]; ok {
			switch sp.useOnTagKeyConflict {
			case "both":
				// If both key and value are conflicting, there's no need to add the collector's tag.
				if cTagVal == sTag.AsString() {
					delete(cTags, sTag.Key)
				}
			case "collector":
				appendSpanTag = false
			case "span":
				// If we choose the span's tag, then remove this conflicting tag from the collector tags.
				delete(cTags, sTag.Key)
			}
		}
		if appendSpanTag {
			procTags = append(procTags, sTag)
		}
	}
	// Append the remaining collector tags.
	for k, v := range cTags {
		procTags = append(procTags, model.String(k, v))
	}
	span.Process.Tags = procTags
}
```

Reply:
I am not good in data-structures but this is definitely lot better and simpler to understand. I decided to use it.